src/MySqlConnector/MySqlClient/Results/ResultSet.cs

```diff
@@ -0,0 +1,358 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using MySql.Data.Protocol.Serialization;
+using MySql.Data.Serialization;
+
+namespace MySql.Data.MySqlClient.Results
+{
+	internal class ResultSet
+	{
+		internal static readonly Task<bool> FalseTask = Task.FromResult(false);
+		internal static readonly Task<bool> TrueTask = Task.FromResult(true);
+
+		public ResultSet(MySqlDataReader dataReader)
+		{
+			DataReader = dataReader;
+		}
+
+		internal async Task<ResultSet> ReadResultSetHeaderAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
```

Comment:
I prefer `ValueTask<T>` (fewer allocations and async state machines on the happy path) to `Task` for APIs that aren't required by ADO.NET.